docs(gcp): Kubeflow cluster upgrade instructions and revamp of env vars (#2372)
kubeflow/website: content/en/docs/gke/deploy/upgrade.md

## Before you start

> To better understand upgrade process, recommend reading the following sections first:

Reviewer: ```suggestion
To better understand upgrade process, you should read the following sections first:
```

- [Understanding the deployment process for management cluster](../management-setup#understanding-the-deployment-process)
- [Understanding the deployment process for Kubeflow cluster](../deploy-cli#understanding-the-deployment-process)

This guide assumes the following settings:

- The `${KF_DIR}` environment variable contains the path to
  your Kubeflow application directory, which holds your Kubeflow configuration
  files. For example, `/opt/my-kubeflow/`.

  ```bash
  KF_DIR=<path to your Kubeflow application directory>
  ```

> - The `${KF_NAME}` environment variable contains the name of your Kubeflow deployment.

Reviewer 2: Define export commands for all environment variables with placeholders for the values. Update the descriptions to indicate the value to replace and what to replace it with.

## Upgrading management cluster

### Upgrade management cluster from v1.1 to v1.2

1. The instructions below assume your current working directory is

   ```bash
   cd "${MGMT_DIR}"
   ```

1. Use your management cluster's kubectl context:

   ```bash
   # Look at all your contexts
   kubectl config get-contexts
   # Select your management cluster's context
   kubectl config use-context "${MGMT_NAME}"
   # Verify the context connects to the cluster properly
   kubectl get namespace
   ```

   If you are using a different enviroment, you can always
   reconfigure the context by:

   ```bash
   make create-context
   ```

> 1. Check your existing config connector version:
>    ```bash

Reviewer: ```suggestion
1. Check your existing config connector version:

   ```bash
```

   # For Kubeflow v1.1, it should be 1.15.1
   $ kubectl get namespace cnrm-system -ojsonpath='{.metadata.annotations.cnrm\.cloud\.google\.com\/version}'
   1.15.1

>    ```

Reviewer: The suggestion I was going to make doesn't render well. Add a blank line after the end of the code block 😄

1. Uninstall the old config connector in the management cluster:

>    ```bash

Reviewer: ```suggestion

   ```bash
```

(Add a blank like before the start of the code block)

   kubectl delete sts,deploy,po,svc,roles,clusterroles,clusterrolebindings --all-namespaces -l cnrm.cloud.google.com/system=true --wait=true
   kubectl delete validatingwebhookconfiguration abandon-on-uninstall.cnrm.cloud.google.com --ignore-not-found --wait=true
   kubectl delete validatingwebhookconfiguration validating-webhook.cnrm.cloud.google.com --ignore-not-found --wait=true
   kubectl delete mutatingwebhookconfiguration mutating-webhook.cnrm.cloud.google.com --ignore-not-found --wait=true

>    ```

Reviewer: The suggestion I was going to make doesn't render well. Add a blank line after the end of the code block 😄

   These commands uninstall the config connector without removing your resources.

> 1. Replace your `./Makefile` with the version in Kubeflow `v1.2.0`: https://github.com/kubeflow/gcp-blueprints/blob/v1.2.0/management/Makefile.

Reviewer: ```suggestion

1. Replace your `./Makefile` with the version in Kubeflow `v1.2.0`: https://github.com/kubeflow/gcp-blueprints/blob/v1.2.0/management/Makefile.
```

   If you made any customizations in `./Makefile`, you should merge your changes with the upstream version. We've refactored the Makefile to move substantial commands into the upstream package, so hopefully future upgrades won't require a manual merge of the Makefile.

1. Update `./upstream/management` package:
   ```bash
   make update

>    ```

Reviewer: The suggestion I was going to make doesn't render well. Add a blank line before and after the code block 👍

1. Use kpt to set user values:
   ```bash
   kpt cfg set -R . name ${NAME}
   kpt cfg set -R . gcloud.core.project ${PROJECT}
   kpt cfg set -R . location ${LOCATION}

>    ```

Reviewer: The suggestion I was going to make doesn't render well. Add a blank line before and after the code block 👍

   Note, you can find out which setters exist in a package and what there current values are by:
   ```bash
   kpt cfg list-setters .

>    ```

Reviewer: The suggestion I was going to make doesn't render well. Add a blank line before and after the code block 👍

1. Apply upgraded config connector:
   ```bash
   make apply-kcc
   ```

>    Note, you can optionally also run `make apply-cluster`, but it should be the same as your existing management cluster.

Reviewer: The suggestion I was going to make doesn't render well. Add a blank line before and after the code block 👍

> 1. Check your config connector upgrade is successful:

Reviewer: ```suggestion

1. Check that your config connector upgrade is successful:
```

   ```bash
   # For Kubeflow v1.2, it should be 1.29.0
   $ kubectl get namespace cnrm-system -ojsonpath='{.metadata.annotations.cnrm\.cloud\.google\.com\/version}'
   1.29.0
   ```

## Upgrading Kubeflow cluster

**DISCLAIMERS**:

- The upgrade process depends on each Kubeflow application to handle the upgrade properly. There's no guarantee on data completeness unless the application provides such a guarantee.

> - Always recommend backing up your data before an upgrade.

Reviewer: ```suggestion
- You should always **back up your data** before an upgrade.
```

WDYT?

Reviewer: Or "you're recommended"... "You" is kind of the point (as in - hey, pay attention please 😸 )

- Upgrading Kubeflow cluster can be a disruptive process, please schedule some downtime and communicate with your users.

> For upgrading from specific versions, there can be manual actions needed, so recommend referring to the specific sections below.

Reviewer: Toward an active voice , e.g.

```suggestion
To upgrade from specific versions of Kubeflow, you may need to take certain manual actions—refer to specific sections in the guidelines below.
```

WDYT?

Author: This is great, thank you!

> Here are general instructions to upgrade Kubeflow:

Reviewer: ```suggestion
General instructions for upgrading Kubeflow:
```

(just a suggestion, of course)

> 1.  The instructions below assume

Reviewer: ```suggestion
1.  The instructions below assume that:
```

>     - your current working directory is

Reviewer: ```suggestion
    - Your current working directory is:
```

      ```bash
      cd ${KF_DIR}
      ```

>     - your kubectl uses a context that connects to your Kubeflow cluster

Reviewer: ```suggestion
    - Your kubectl uses a context that connects to your Kubeflow cluster:
    
```

      ```bash
      # List your existing contexts
      kubectl config get-contexts
      # Use the context that connects to your Kubeflow cluster
      kubectl config use-context ${KF_NAME}
      ```

1.  Edit the Makefile at `./Makefile` and change `MANIFESTS_URL` to point at the version of Kubeflow manifests you want to use

    - Refer to the [kpt docs](https://googlecontainertools.github.io/kpt/reference/pkg/) for more info about supported dependencies

1.  Update the local copies:

    ```bash
    make update
    ```

1.  Redeploy:

    ```bash
    make apply
    ```

    To evaluate the changes before deploying them you can run `make hydrate` and then compare the contents

>     of `.build` to what is currently deployed.

Reviewer 3: > compare the contents of `.build` to what is currently deployed.

maybe state clear where/how to check what is currently deployed.?

Author: Good point, reworded

### Upgrade Kubeflow cluster from v1.1 to v1.2

> 1.  The instructions below assume

Reviewer: ```suggestion
1.  The instructions assume that:
```

Author: I think we need to keep this `below`, because in some docs, this sentence appears multiple times.

>     - your current working directory is

Reviewer: ```suggestion
    - Your current working directory is:
```

      ```bash
      cd ${KF_DIR}
      ```

>     - your kubectl uses a context that connects to your Kubeflow cluster

Reviewer: ```suggestion
    - Your kubectl uses a context that connects to your Kubeflow cluster:
```

      ```bash
      # List your existing contexts
      kubectl config get-contexts
      # Use the context that connects to your Kubeflow cluster
      kubectl config use-context ${KF_NAME}
      ```

> 2.  (Recommended) Replace your `./Makefile` with the version in Kubeflow `v1.2.0`: https://github.com/kubeflow/gcp-blueprints/blob/v1.2.0/kubeflow/Makefile.

Reviewer 3: Just curious, why do we start staring using sequential numbers for the list items instead of all `1.`?

Author: My editor sometimes does that automatically, thanks for the correction!

    If you made any customizations in `./Makefile`, you should merge your changes with the upstream version.

    This step is recommended, because we introduced usability improvements and fixed compatibility for newer Kustomize versions (while still being compatible with Kustomize v3.2.1) to the Makefile. However, the deployment process is backward-compatible, so this is recommended, but not required.

3.  Update `./upstream/manifests` package:
    ```bash
    make update
    ```
4.  Before applying new resources, you need to delete some immutable resources that were updated in this release:

    ```bash
    kubectl delete statefulset kfserving-controller-manager -n kubeflow --wait
    kubectl delete crds experiments.kubeflow.org suggestions.kubeflow.org trials.kubeflow.org
    ```

>     **WARNING**: this deletes all Katib running resources.

Reviewer: ```suggestion
    **WARNING**: This step **deletes** all Katib running resources.
```
